Support Nullable(JSON) in ClickHouse

aldofunes opened the pull request:
This is my attempt at addressing issue #3546.

I managed to make it work on my setup, which pulls NULL JSONB columns in a Postgres db into Nullabl(JSON) columns in ClickHouse

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PeerDB-io/peerdb?pullRequest=3626) <br/>All committers have signed the CLA.

aldofunes commented:
@jgao54, you are totally right, how did I miss that?

I found that in the clone step, the avro files have a different schema from the CDC step.

Here's an example:

`java -jar avro-tools-1.12.1.jar getschema cdc_file.avro`
```json
{
  "type" : "record",
  "name" : "_peerdb_raw_aggregators_basculin_dwh",
  "fields" : [ {
    "name" : "_peerdb_uid",
    "type" : "string"
  }, {
    "name" : "_peerdb_timestamp",
    "type" : "long"
  }, {
    "name" : "_peerdb_destination_table_name",
    "type" : "string"
  }, {
    "name" : "_peerdb_data",
    "type" : "string"
  }, {
    "name" : "_peerdb_record_type",
    "type" : [ "null", "long" ]
  }, {
    "name" : "_peerdb_match_data",
    "type" : [ "null", "string" ]
  }, {
    "name" : "_peerdb_batch_id",
    "type" : [ "null", "long" ]
  }, {
    "name" : "_peerdb_unchanged_toast_columns",
    "type" : [ "null", "string" ]
  } ]
}
```

`java -jar avro-tools-1.12.1.jar getschema clone_file.avro`
```json
{
  "type" : "record",
  "name" : "aggregators_basculin_activity_views",
  "fields" : [ {
    "name" : "user_id_0",
    "type" : {
      "type" : "string",
      "logicalType" : "uuid"
    }
  }, {
    "name" : "operation_id_1",
    "type" : {
      "type" : "string",
      "logicalType" : "uuid"
    }
  }, {
    "name" : "data_2",
    "type" : [ "null", "string" ]
  }, {
    "name" : "created_at_3",
    "type" : {
      "type" : "long",
      "logicalType" : "timestamp-micros"
    }
  }, {
    "name" : "updated_at_4",
    "type" : {
      "type" : "long",
      "logicalType" : "timestamp-micros"
    }
  } ]
}
```

In other words, the CDC step is wrapping the record in an envelope, while the Clone step is dumping the record by itself.

For the life of me, I could not get the tests to run on my machine, but I was able to figure out that in the CDC step, it needs to first extract the `jsonb` column to a `Nullable(String)` and then cast that to a `Nullable(JSON)`, using `JSONExtractString()` does not work because `null` values get converted to an empty string (instead of `null`) and hence, the cast to `Nullable(JSON)` fails.

I basically had to split the handling of both `JSON` and `Nullable(JSON)` as separate steps. I have zero experience with golang, so please forgive me if my approach is too naive.

jgao54 commented on flow/connectors/clickhouse/normalize_query.go:
```diff
@@ -193,7 +193,7 @@ func (t *NormalizeQueryGenerator) BuildQuery(ctx context.Context) (string, error
 				)
 			}
-		case "JSON", "Nullable(JSON)":
+		case "JSON":
```
Can be combined into:

```suggestion
		case "JSON", "Nullable(JSON)":
			stringType := strings.Replace(clickHouseType, "JSON", "String", 1)
			fmt.Fprintf(&projection,
				"JSONExtract(_peerdb_data, %s, '%s')::%s AS %s,",
				peerdb_clickhouse.QuoteLiteral(colName),
				stringType,
				clickHouseType,
				peerdb_clickhouse.QuoteIdentifier(dstColName),
			)
			if t.enablePrimaryUpdate {
				fmt.Fprintf(&projectionUpdate,
					"JSONExtract(_peerdb_match_data, %s, '%s')::%s AS %s,",
					peerdb_clickhouse.QuoteLiteral(colName),
					stringType,
					clickHouseType,
					peerdb_clickhouse.QuoteIdentifier(dstColName),
				)
			}

```

aldofunes committed:
fix(clickhouse): handle nullable JSON columns in normalization

Add support for `Nullable(JSON)` type in ClickHouse connector by:
- Recognizing `Nullable(JSON)` in schema type mapping
- Using `JSONExtract` with proper type casting for both JSON and
  Nullable(JSON) columns
- Casting nullable JSON fields to `Nullable(JSON)` in table functions
